Cylindrical algebraic decomposition implementation

mmaaz-git opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### Brief description of what is fixed or changed

This is an implementation of cylindrical algebraic decomposition (CAD) in SymPy, which enables the solution of systems of multivariate polynomial (in)equalities. Example:

```
>>> from sympy.abc import x,y
>>> from sympy.solvers import polysys
>>> 
>>> polysys.solve_poly_system_cad([-x**2-1>0], [x])
[]
>>> polysys.solve_poly_system_cad([x*y>2, x<0.1, x**2+y > 3], [x,y])
[{y: -2, x: -sqrt(5) - 1}]
```

The two main functions that expose this functionality are (1) `cylindrical_algebraic_decomposition()`, which computes a CAD _adapted_ to a set of polynomials (a decomposition into cells over which each polynomial is sign-invariant); and (2) `solve_poly_system_cad()`, which takes a set of polynomial (in)equalities, constructs polynomial atoms (by subtracting the right-hand side), then passing it into `cylindrical_algebraic_decomposition()` and lastly checking each cell for the system to hold. 

Technically, this required the implementation of: computing reducta (`red()` and `red_set()`), computing subresultant coefficients (`subresultant_polynomials()` and `subresultant_coefficients()`), and the Hong projector operator (`projone()`, `projtwo()`, and `hongproj()`). With these functions, implementing CAD is rather straightforward. 

#### Technical details

For posterity's sake, I will summarize the technical details of CAD, as the literature is somewhat difficult to parse together (at least, it was for me, as it took me many months to really understand). Hopefully, this enables others to work on CAD as well as confirm that my code is correct.

CAD takes a set of k-variate polynomials, repeatedly projects them down to (k-1)-space, (k-2)-space, and so on, up to 1-space. At each step, a set of polynomials are generated called projection factors. In 1-space, the roots of the projection factors at that level are computed and ordered. Then, _cells_ are constructed as follows: the cells are the roots themselves, the intervals between consecutive roots, and the ray from $-\infty$ to the smallest root, and the ray from the largest root to $+\infty$. In each cell, we compute sample points: if the cell is a point, the sample point is the point itself; for the finite intervals, we take the average of the endpoints, and for the rays we take the smallest root minus one and the largest root plus one.

We continue this procedure for 2-space, 3-space, and so on, back up to k-space. At each level, we loop over the sample points already constructed, and plug them into the projection factors at that level -- they are now univariate polynomials, so we can compute their roots and construct cells, as described before. Of course, you can see that this process can blow up quite fast, as over each cell in 1-space, we construct a number of cells in 2-space; then over each cell in 2-space, we construct a number of cells in 3-space, and so on. This process is pretty mechanical: just requires looping and computing roots.

The main question is how to actually construct these projection factors. Here we use the Hong projection operator: see [Hong 1990](https://dl.acm.org/doi/pdf/10.1145/96877.96943). Note there are some newer operators now, like the Brown-McCallum, but they require the polynomials to obey some properties, whereas the Hong one always works. This could be a change in a future implementation. Given a set of multivariate polynomials $F$, we choose one of the variables as the main variable, or "mvar". The Hong projection operator $PROJH$ is defined as:

$$ PROJH(F) = PROJ_1(F) \cup PROJ_2(F) $$

$$ PROJ_1(F) = \bigcup_{f \in F, f' \in RED(f)} \left( LC(f') \cup PSC(f', D(f')) \right) $$

$$ PROJ_2(F) = \bigcup_{f,g \in F, f \lt g} \quad \bigcup_{f' \in RED(f)} PSC(f', g) $$

Note that in Hong's paper, what we call $PROJ_2$ here is labelled $PROJ^*_2$ (he was showing an improvement over Collins' original operator). This seems quite complex but really it's a bunch of loops. Above, $LC$ is the leading coefficient, wrt to the mvar, $D$ is the derivative operator, and $\lt$ represents an arbitrary "linear" ordering so as to not loop over redundant pairs. 

$RED$ represents the _reducta set_. For a function $f$, its reductum is $f - LT(f)$, where $LT$ is the leading term. The $i$th level reducta, denoted $red^i$ is defined recursively: $red^0(f) = f$ and $red^{i+1}(f) = red(red^i(f))$. Then, the set of reducta, $RED(f) = \\{red^i(f), 0 \le i \le deg(f)\\}$. I have written functions `red()` and `red_set()` for these. In the Hong paper, they exclude $0$ from $RED$ by definition. I keep it in in case someone needs to access the $i$-th level reducta for some reason, so that the length is always $deg(f)+1$. Anyways we ignore constants in the projection factor set later.

$PSC$ refers to the _principal subresultant coefficients_. In order to compute these, we need the subresultants themselves. In SymPy, the `subresultants()` function returns the subresultant polynomial remainder sequence (PRS). We can compute the subresultants, also known as the subresultant polynomials, from the remainders in the PRS (although this took me a long time to figure it out...). 

Suppose we have two polynomials $f$ and $g$, and without loss of generality $deg(f) \gt deg(g)$. There will be $deg(g)+1$ subresultants, which we will initialize as $0$. Assume we have the remainders from the PRS.  Then, the remainder $r_i$ is the $(deg(r_{i-1} - 1)$-th subresultant. Additionally, if $deg(r_i) < deg(r_{i-1}) - 1$, then the $deg(r_i)$-th subresultant is $r_i \cdot LC(r_i)^{deg(r_{i-1}) - deg(r_i) - 1}$ . So, a remainder can possibly "contribute" to two subresultants. I don't fully understand subresultant PRS, so would greatly appreciate if someone could confirm this. Lastly, given the subresultants in increasing degree order, the $i$-th subresultant coefficient is the coefficient of $x^i$ in the $i$-th subresultant. These are implemented in the functions `subresultant_polynomials()` and `subresultant_coefficients()`.

With all of this notation defined, it's easy to implement the projection operators: just loop and call these operators! The implementations are in `projone()`, `projtwo()`, and `hongproj()`. The function `hongproj()` just puts them together, and does some cleaning up, namely removing constant numbers, and makes polynomials unique up to sign, as having both $f$ and $-f$ in a projection factor set does not affect the final CAD construction.

A pesky computational issue is how to compute and store the roots. I wrote a helper function `get_nice_roots()` which basically tries to keep roots algebraic if possible, but evaluates `CRootOf` objects numerically. Also, if a polynomial cannot be solved algebraically (e.g. with irrational coefficients), then it numerically finds the roots. This is I think the best that can be done right now with SymPy.

Once we have a CAD adapted to a set of polynomials, we can check if inequalities hold of the form $f \bowtie 0$, where $\bowtie \in \\{ \lt, \gt, \le, \ge, =, \neq \\}$ just by testing the sample points, as they represent cells over which the polynomials are sign-invariant. Thus, we can show that the system is inconsistent if none of the sample points we built satisfy the system; or we can find satisfying points. Once we have a CAD, solving a polynomial system becomes a simple search (although over possibly a very large set!).


#### Implementation overview

Line 437 in polysys.py contains the `solve_poly_system_cad()` function. Line 500 and beyond are all the "background" functions. 
- `solve_poly_system_cad(ineqs, gens, return_one_sample=True)` runs CAD on a set of (in)equalities. It can return one sample point, or sample points from all cells; if the system is inconsistent it will return `[]`.
- `red(f, mvar)`: reductum of a function wrt mvar
- `red_set(f, mvar)`: reducta set ($RED$) of a function wrt mvar
- `subresultant_polynomials(f, g, mvar)`: subresultants of $f$ and $g$ wrt mvar
- `subresultant_coefficients(f, g, mvar)`: subresultant coefficients (what Hong's paper calls $PSC$) of $f$ and $g$ wrt mvar
- `get_nice_roots(poly)`: computes roots algebraically if possible, otherwise numeric; not meant to be public function (?)
- `projone(F, mvar)`, `projtwo(F, mvar)`, `hongproj(F, mvar)`: Hong projector operator
- `cylindrical_algebraic_decomposition(F, gens)`: computes the CAD

#### Other comments

For now, I have added everything into solvers/polysys.py, because there was already a function called `solve_poly_system()`. However, I'm totally open to suggestions on how to organize the many functions better. For example, the functions related to reducta and subresultants could probably be put alongside the existing subresultant/resultant functions -- I'm not sure where they are situated. Also, since `projone()` and `projtwo()` are very specific to the Hong operator, perhaps they should not be standalone functions, and instead functions defined within `hongproj()`? Lastly, perhaps the main public-facing API should be part of `solve_poly_system()`, with some additional logic in there to call the CAD function if needed. I am very open to suggestions about these things from people more familiar with the codebase!


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* solvers
    * Added cylindrical algebraic decomposition (CAD) as a solver for multivariate polynomial (in)equalities.

<!-- END RELEASE NOTES -->

mmaaz-git committed:
Initial addition of CAD to polysys

Added functions for reducta and principal subresultant coefficients. Implemented Hong projection operator. Implemented cylindrical algebraic decomposition, which makes a CAD adapted to a set of polynomials. Lastly, made solve_poly_system_cad, which makes the CAD, then evalutes the truth using sample cells.

mmaaz-git committed:
Tests for the reducta functions

Wrote tests for red(), red_level(), and red_set(). Began to write tests for PSC function, but I encountered an error. I have to write a subresultants function from scratch.

mmaaz-git committed:
Fixed subresultant coefficients

Better understood what subresultants() already does in SymPy. Wrote functions for subresultant polynomials and subresultant coefficients, to match Mathematica convention of degree sequence. The subresultant coefficients is what we actually need for CAD.

mmaaz-git committed:
Fixed subresultant coefficients hopefully

Realized after testing that subresultants had to be normalized by a certain factor, so did that now. Also, generated a diverse set of examples and used Mathematica to get the subresultants, to write unit tests for our function.

mmaaz-git committed:
Fixing code formatting and tests

mmaaz-git committed:
Fixed subresultants again

Found it failed in some multivariate cases. I think it is fixed now, after spending lots of time reading. I tested it with a bunch of bivariate examples. It would be good to get this checked though. Also added bivariate tests for it.

mmaaz-git committed:
Better handle roots and cleaned projection facto

Had issues with convergence of root calculations. Solved it by adding factoring before the root computations as the convergence issue was just due to large coefficients. Also cleaned up the Hong projection set by removing constants and removing polynomials that are a sign away from another one. I think we can possibly remove those that are a constant away, but not sure!

mmaaz-git committed:
Added tests to projections and cad construction

Added univar and multivar tests to projone, projtwo, hongproj, cad function, and solve_poly_system_cad. Fixed some minor issues in these functions too.

mmaaz-git committed:
Updated documentation

Updated docstrings for the CAD related functions, and added doctests.

mmaaz-git committed:
Updates to testing, fixed code quality

Mistakenly omitted some unit tests, but put them back in now. Minor code clean up to pass code quality and flake8.

mmaaz-git committed:
Update test_polysys.py

Removed use of sympify in writing test solutions.

mmaaz-git committed:
Update test_polysys.py

Removed sympify import

mmaaz-git committed:
Update .mailmap

mmaaz-git committed:
Fixed backslashes in docstrings in polysys.py

mmaaz-git committed:
Update .mailmap

Had to add someone to make the tests go through.

mmaaz-git committed:
Merge branch 'master' into cad-branch

mmaaz-git committed:
Update .mailmap

mmaaz-git committed:
Merge branch 'cad-branch' of https://github.com/mmaaz-git/sympy into cad-branch

oscarbenjamin commented on sympy/solvers/polysys.py:
```diff
@@ -429,3 +432,567 @@ def solve_triangulated(polys, *gens, **args):
+        if curr_factor.is_number:
+            continue
+
+        try:
+            new_roots = real_roots(curr_factor)
+            for i, r in enumerate(new_roots):
+                # want to avoid CRootOf
+                if r.has(ComplexRootOf):
+                    new_roots[i] = r.evalf()
+        except NotImplementedError:
+            new_roots = [root for root in nroots(curr_factor) if root.is_real]
```
I don't think we can get away with using numeric approximations here. If we are to handle the 0D cells then we need to do it exactly. I assume that is why the Eq case here fails:
```python
In [11]: solve_poly_system_cad([x**5 - x + 1 > 0], [x])
Out[11]: [{x: -0.167303978261419}]

In [12]: solve_poly_system_cad([x**5 - x + 1 < 0], [x])
Out[12]: [{x: -2.16730397826142}]

In [13]: solve_poly_system_cad([Eq(x**5 - x + 1, 0)], [x])
Out[13]: []
```
Also there is no reason why the outputs cannot be exact rational numbers for the inequality cases. In this particular example they could even be integers.

I think we really need the implementation to be based entirely on rational numbers and RootOf rather than radicals and floats. The `nroots` function is not reliable in general and neither is replacing RootOf with floats.

oscarbenjamin commented on sympy/solvers/polysys.py:
```diff
@@ -429,3 +432,567 @@ def solve_triangulated(polys, *gens, **args):
+    # prepare the atoms
+    atoms = [Poly(p.lhs - p.rhs, gens) for p in seq]
+
+    sample_points = cylindrical_algebraic_decomposition(atoms, gens)
+    valid_samples = []
+
+    for sample in sample_points:
+        if all(expr.subs(sample) for expr in seq):
+            valid_samples.append(sample)
```
There are a couple of problems with this approach.

Firstly the sample points can contain floats and so this depends on testing exact equality of floats which does not generally work. Really floats should not be used.

Secondly this is trying to infer after the fact what the sign was of each polynomial when this information is already known during each lifting step when computing the cylindrical algebraic decomposition. It would be better to keep track of the signs of the polynomials during lifting.

oscarbenjamin commented:
Thanks for opening this. This is good work and we definitely want something like this in SymPy. A lot of changes and further discussion would be needed before it is ready though.

For now I just want to focus on the highest-level point which is that the algorithm here uses floats in a way that is unreliable and it therefore cannot handle the lower-dimensional cells correctly. That is why in gh-26177 we focused only on the full-dimensional cells and using exact rational numbers for all sample points.

> A pesky computational issue is how to compute and store the roots. I wrote a helper function `get_nice_roots()` which basically tries to keep roots algebraic if possible, but evaluates `CRootOf` objects numerically. Also, if a polynomial cannot be solved algebraically (e.g. with irrational coefficients), then it numerically finds the roots. This is I think the best that can be done right now with SymPy.

Firstly we should use RootOf where at all possible. It gives a unified representation of all algebraic numbers. Converting to radicals can be done at the end or by the user but the internal implementation should try to use only rational numbers and RootOf. (Likewise the internal implementation should use Poly everywhere rather than mixing Poly and Expr.)

Secondly when you say "this is the best that can be done right now with SymPy" I think we should consider instead what else in SymPy should be improved to make this work. Using `nroots` or `evalf` will not work in general.

oscarbenjamin commented:
> if a polynomial cannot be solved algebraically (e.g. with irrational coefficients)

Take a look at https://github.com/sympy/sympy/issues/22943#issuecomment-2220562152

mmaaz-git commented:
Thanks for the quick review! I agree with all your points. I'll implement a function that obtains a sample points in an open interval that is an integer, if possible, or else a rational. 

As for the 0D cells, my concern is whether SymPy is able to find the roots of a polynomial with coefficients that are CRootOf objects. This may happen during the lifting stage. For example,

```
>>> from sympy.abc import x,y,z
>>> from sympy import solve, CRootOf, Poly
>>> 
>>> sample = {y : CRootOf(y**3 + y**2 - 1, 0), z : CRootOf(z**5 + z**2 - 1, 0)}
>>> poly = y * x + 3 * y ** 2 - z + x**2 - x**5
>>> 
>>> sub = poly.subs(sample)
>>> 
>>> sub = Poly(sub, x)
>>> 
>>> print(sub)
Poly(-x**5 + x**2 + CRootOf(y**3 + y**2 - 1, 0)*x - CRootOf(z**5 + z**2 - 1, 0) + 3*CRootOf(y**3 + y**2 - 1, 0)**2, x, domain='ZZ[CRootOf(y**3 + y**2 - 1, 0),CRootOf(z**5 + z**2 - 1, 0)]')
>>> 
>>> print(sub.real_roots())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/maaz/Documents/sympy/sympy/polys/polytools.py", line 3638, in real_roots
    reals = sympy.polys.rootoftools.CRootOf.real_roots(f, radicals=radicals)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/maaz/Documents/sympy/sympy/polys/rootoftools.py", line 412, in real_roots
    return cls._get_roots("_real_roots", poly, radicals)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/maaz/Documents/sympy/sympy/polys/rootoftools.py", line 776, in _get_roots
    coeff, poly = cls._preprocess_roots(poly)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/maaz/Documents/sympy/sympy/polys/rootoftools.py", line 744, in _preprocess_roots
    raise NotImplementedError(
NotImplementedError: sorted roots not supported over ZZ[CRootOf(y**3 + y**2 - 1, 0),CRootOf(z**5 + z**2 - 1, 0)]
>>> 
>>> print(solve(sub))
[]
>>> 
>>> print(sub.nroots())
[1.28678114844068, -0.650871114532895 - 0.611479884110014*I, -0.650871114532895 + 0.611479884110014*I, 0.00748054031255526 - 0.936848109106606*I, 0.00748054031255526 + 0.936848109106606*I]
```

Here, `.real_roots()` fails, `solve()` gives an empty list, but `.nroots()` gives a single real root. Is there any way to do this? Is this related to the comment you linked to above? I would definitely be interested in integrating such functionality so that we can handle everything properly during the CAD construction.

oscarbenjamin commented:
> Here, `.real_roots()` fails, `solve()` gives an empty list, but `.nroots()` gives a single real root. Is there any way to do this?

See the linked code at https://github.com/sympy/sympy/issues/22943#issuecomment-2220562152

It should be able to do this although there does seem to bugs in other places highlighted by your example:
```python
In [19]: poly.subs(sample)
Out[19]: 
                                                                                           2
   5    2            ⎛ 3    2       ⎞          ⎛ 5    2       ⎞            ⎛ 3    2       ⎞ 
- x  + x  + x⋅CRootOf⎝y  + y  - 1, 0⎠ - CRootOf⎝z  + z  - 1, 0⎠ + 3⋅CRootOf⎝y  + y  - 1, 0⎠ 

In [20]: poly.subs(sample).as_poly(x)
Out[20]: 
Poly(-x**5 + x**2 + CRootOf(y**3 + y**2 - 1, 0)*x - CRootOf(z**5 + z**2 - 1, 0) + 3*CRootOf(y**3 +  ↪

↪ y**2 - 1, 0)**2, x, domain='ZZ[CRootOf(y**3 + y**2 - 1, 0),CRootOf(z**5 + z**2 - 1, 0)]')

In [21]: poly.subs(sample).as_poly(x, extension=True)
Out[21]: 
Poly(-x**5 + x**2 + CRootOf(y**3 + y**2 - 1, 0)*x - CRootOf(z**5 + z**2 - 1, 0) + 3*CRootOf(y**3 +  ↪

↪ y**2 - 1, 0)**2, x, domain='ZZ[CRootOf(y**3 + y**2 - 1, 0),CRootOf(z**5 + z**2 - 1, 0)]')
```
That last example should have constructed an algebraic field. I'm not sure why `construct_domain` fails:
```python
In [29]: construct_domain([CRootOf(y**3 + y**2 - 1, 0),CRootOf(z**5 + z**2 - 1, 0)], extension=True)
Out[29]: 
(ZZ[CRootOf(y**3 + y**2 - 1, 0),CRootOf(z**5 + z**2 - 1, 0)],
 [CRootOf(y**3 + y**2 - 1, 0), CRootOf(z**5 + z**2 - 1, 0)])
```
This fix is needed:
```diff
diff --git a/sympy/polys/rootoftools.py b/sympy/polys/rootoftools.py
index cfd0af01cb..5bcba1a836 100644
--- a/sympy/polys/rootoftools.py
+++ b/sympy/polys/rootoftools.py
@@ -298,6 +298,7 @@ class ComplexRootOf(RootOf):
     is_complex = True
     is_number = True
     is_finite = True
+    is_algebraic = True
 
     def __new__(cls, f, x, index=None, radicals=False, expand=True):
         """ Construct an indexed complex root of a polynomial.
```
That gives
```python
In [14]: construct_domain([CRootOf(y**3 + y**2 - 1, 0),CRootOf(z**5 + z**2 - 1, 0)], extension=True)
Out[14]: 
(QQ<CRootOf(y**3 + y**2 - 1, 0) + CRootOf(z**5 + z**2 - 1, 0)>,
 [ANP([559667376622/2126465723061, 2651238493942/2126465723061, 4867605853559/2126465723061, 3035219759674/2126465723061, -938760822595/708821907687, -5708672543009/2126465723061, 3891525098632/708821907687, 36003892562578/2126465723061, -25044001377814/2126465723061, -136182147762611/2126465723061, -152460905408198/2126465723061, -142709969313743/2126465723061, -70105632326617/2126465723061, -22192406036605/2126465723061, -358996135043/236273969229], [1, 5, 10, 8, -3, -11, 18, 69, -27, -251, -337, -341, -215, -96, -26, -5], QQ),
  ANP([-559667376622/2126465723061, -2651238493942/2126465723061, -4867605853559/2126465723061, -3035219759674/2126465723061, 938760822595/708821907687, 5708672543009/2126465723061, -3891525098632/708821907687, -36003892562578/2126465723061, 25044001377814/2126465723061, 136182147762611/2126465723061, 152460905408198/2126465723061, 142709969313743/2126465723061, 70105632326617/2126465723061, 24318871759666/2126465723061, 358996135043/236273969229], [1, 5, 10, 8, -3, -11, 18, 69, -27, -251, -337, -341, -215, -96, -26, -5], QQ)])
```
Other problems remain though:
```python
In [23]: p = sub.as_poly(x, extension=True)

In [24]: p.lift()
...
CoercionFailed: Cannot convert ANP([25814020042/708821907687, 102612150188/2126465723061, -370944676484/2126465723061, -949257895736/2126465723061, -114117277984/708821907687, 958120754932/2126465723061, 2714573351182/2126465723061, -751318831508/708821907687, -15131823204712/2126465723061, 1825624657862/708821907687, 28507439262118/2126465723061, 1860082400012/2126465723061, 20503572026260/2126465723061, 2913384869830/2126465723061, 397586647142/2126465723061], [1, 5, 10, 8, -3, -11, 18, 69, -27, -251, -337, -341, -215, -96, -26, -5], QQ) of type <class 'sympy.polys.polyclasses.ANP'> from QQ<CRootOf(y**3 + y**2 - 1, 0) + CRootOf(z**5 + z**2 - 1, 0)> to QQ
```
I'm not sure what `p.lift` is doing there but it should be possible to handle that as a resultant calculation I think. The polynomial p is:
```python
In [29]: p.rep
Out[29]: DMP_Python([ANP([-1], [1, 5, 10, 8, -3, -11, 18, 69, -27, -251, -337, -341, -215, -96, -26, -5], QQ), ANP([], [1, 5, 10, 8, -3, -11, 18, 69, -27, -251, -337, -341, -215, -96, -26, -5], QQ), ANP([], [1, 5, 10, 8, -3, -11, 18, 69, -27, -251, -337, -341, -215, -96, -26, -5], QQ), ANP([1], [1, 5, 10, 8, -3, -11, 18, 69, -27, -251, -337, -341, -215, -96, -26, -5], QQ), ANP([559667376622/2126465723061, 2651238493942/2126465723061, 4867605853559/2126465723061, 3035219759674/2126465723061, -938760822595/708821907687, -5708672543009/2126465723061, 3891525098632/708821907687, 36003892562578/2126465723061, -25044001377814/2126465723061, -136182147762611/2126465723061, -152460905408198/2126465723061, -142709969313743/2126465723061, -70105632326617/2126465723061, -22192406036605/2126465723061, -358996135043/236273969229], [1, 5, 10, 8, -3, -11, 18, 69, -27, -251, -337, -341, -215, -96, -26, -5], QQ), ANP([443504286433/2126465723061, 2497320268660/2126465723061, 5424022868285/2126465723061, 4459106603278/2126465723061, -767584905619/708821907687, -7145853675407/2126465723061, 2534238423041/708821907687, 39384827304364/2126465723061, -2346266570746/2126465723061, -144397458722990/2126465723061, -195222064301375/2126465723061, -145500092913761/2126465723061, -100860990366007/2126465723061, -28688949064411/2126465723061, -1275781728700/708821907687], [1, 5, 10, 8, -3, -11, 18, 69, -27, -251, -337, -341, -215, -96, -26, -5], QQ)], QQ<CRootOf(y**3 + y**2 - 1, 0) + CRootOf(z**5 + z**2 - 1, 0)>)
```

mmaaz-git commented:
Thanks. I had a look at your code as well as the related PR #23077 where they use resultants. I'm not sure what is going on in `.lift()` either, so I replaced the call to `.lift()` with code from that PR, which I put in a function called `algebraic_to_rational()`.

I also did notice that `.count_roots()` fails sometimes for algebraic coefficients. Here is an example.

```python
from sympy.abc import x
from sympy import Poly, sqrt

poly = Poly(sqrt(2) * x**5 - sqrt(3) * x**3 + 1, x)
print(poly.count_roots())
print(poly.nroots())
```
Which gives:

```
3
[-1.25721399074364, -0.317589667629025 - 0.683828065470047*I, -0.317589667629025 + 0.683828065470047*I, 0.946196663000843 - 0.306723225079745*I, 0.946196663000843 + 0.306723225079745*I]
```

So `.count_roots()` says 3 roots but numerically only one real root is there. For now, I have thus replaced it with using `.nroots()`.

Here is the full implementation with some examples. As you point out, it can be a bit slow especially when the rational lift (is that the right term?) gets quite complicated.

```python
from sympy import Poly, PurePoly, sqrt, CRootOf, QQ, ZZ, symbols, re, im
from sympy.abc import x,y,z
from sympy.polys.numberfields import primitive_element
from sympy.polys.polytools import resultant

def real_roots_alg(p, x):
    """Real roots of a polynomial with algebraic coefficients.

    >>> real_roots_alg(x**2 + 2*sqrt(2)*x + 1, x)
    [-sqrt(2) - 1, 1 - sqrt(2)]
    >>> for r in real_roots_alg(x**5 - x + sqrt(2), x):
    ...     print(r, ':', r.evalf(3))
    CRootOf(x**10 - 2*x**6 + x**2 - 2, 0) : -1.21

    """
    p_Alg = Poly(p, x, extension=True)

    if p_Alg.domain in (QQ, ZZ):
        return p_Alg.real_roots()

    return _get_roots_alg(p_Alg, real=True)


def all_roots_alg(p, x):
    """All roots of a polynomial with algebraic coefficients.

    >>> all_roots_alg(x**2 + sqrt(2)*x + 1, x)
    [-sqrt(2)/2 - sqrt(2)*I/2, -sqrt(2)/2 + sqrt(2)*I/2]
    >>> for r in all_roots_alg(x**3 + sqrt(2)*x + 1, x):
    ...     print(r, ':', r.evalf(3))
    CRootOf(x**6 + 2*x**3 - 2*x**2 + 1, 1) : -0.574
    CRootOf(x**6 + 2*x**3 - 2*x**2 + 1, 4) : 0.287 - 1.29*I
    CRootOf(x**6 + 2*x**3 - 2*x**2 + 1, 5) : 0.287 + 1.29*I

    """
    p_Alg = Poly(p, x, extension=True)

    if p_Alg.domain == QQ:
        return p_Alg.all_roots()

    assert p_Alg.domain.is_AlgebraicField

    return _get_roots_alg(p_Alg, real=False)


def _get_roots_alg(p_Alg, *, real):
    """Get roots of a polynomial with algebraic coefficients."""

    _, p_Alg_factors = p_Alg.sqf_list()

    roots_mult = {}

    for p_Alg_f, m1 in p_Alg.sqf_list()[1]:
        for r in _get_roots_alg_sqf(p_Alg_f, real=real):
            roots_mult[r] = m1

    assert len(set(roots_mult)) == len(roots_mult)

    roots_list = []

    for r in sorted_roots(roots_mult):
        roots_list.extend([r] * roots_mult[r])

    return roots_list


def algebraic_to_rational(poly, x):
    """Convert a polynomial with algebraic coefficients to one with rational coefficients."""
    coeffs = poly.all_coeffs()
    dummy = symbols('dum')
    A, r_A, B_s = primitive_element(coeffs, dummy, ex=True)
    B = sum(Poly(Bi, dummy).as_expr() * x**i for i, Bi in enumerate(B_s[::-1]))
    rat_poly = PurePoly(resultant(A, B, dummy), x, domain='QQ')
    return rat_poly


def _get_roots_alg_sqf(p_Alg, *, real):
    """Get roots of a square-free polynomial with algebraic coefficients."""

    p_QQ = algebraic_to_rational(p_Alg, x)

    assert p_QQ.domain == QQ

    if real:
        # XXX: Not sure how well count_roots works with algebraic coefficients.
        #num_roots = p_Alg.count_roots()
        num_roots = len([r for r in p_Alg.nroots() if r.is_real])
        candidates = p_QQ.real_roots()
    else:
        num_roots = p_Alg.degree()
        candidates = p_QQ.all_roots()

    prec = 10

    while len(candidates) > num_roots:
        candidates = [c for c in candidates if _possible_root(p_Alg, c, prec)]
        prec *= 2

    assert len(candidates) == num_roots

    return candidates


def _possible_root(p_Alg, r, prec):
    """Check if r is a possible root of p_Alg."""
    # XXX: In principle we could use count_roots and the interval for r here
    # but count_roots with an interval does not seem to work for polynomials
    # with algebraic coefficients.
    #
    #  >>> nroots(x**2 + 2*sqrt(2)*x + 1)
    #  [-2.4142135623731, -0.414213562373095]
    #  >>> Poly(x**2 + 2*sqrt(2)*x + 1, extension=True).count_roots()
    #  2
    #  >>> Poly(x**2 + 2*sqrt(2)*x + 1, extension=True).count_roots(-10, 10)
    #  0
    #
    r_f = p_Alg(r).evalf(prec, maxn=2*prec)
    possible_zero = abs(r_f)._prec < 2
    return possible_zero


def sorted_roots(roots):
    """Sort roots according to RootOf ordering."""
    #
    # Sorting with key (re(r), -im(r)) will fail if multiple roots have the
    # same real part so work only with one root from each complex conjugate
    # pair.
    #
    real_roots = [r for r in roots if r.is_real]
    complex_roots = [r for r in roots if not r.is_real and im(r) < 0]

    real_roots.sort()
    complex_roots.sort(key=lambda r: (re(r), -im(r)))

    all_roots = real_roots
    for cr in complex_roots:
        all_roots.append(cr)
        all_roots.append(cr.conjugate())

    return all_roots


# Some examples

for poly in [
    Poly(sqrt(3) * x**5 + x**3 - 1, x),
    Poly(sqrt(2) * x**5 - sqrt(3) * x**3 + 1, x), # count_roots() fails on this
    Poly(CRootOf(y**5 + y**3 - 1, 0) ** 2 + x**5 - 3, x),
    Poly(sqrt(3) * x**2 + x**5 - CRootOf(y**5 + y**3 - 1, 0) ** 2 * x - 1, x),
    Poly(CRootOf(z**5 + z**3 - 1, 0)*x + CRootOf(y**5 + y**3 - 3, 0) * x**2 - 1, x)
]:
    print(poly)

    # lift to rational to get the roots
    print(real_roots_alg(poly, x))

    # numerically compute the roots from the rational lift
    print([r.evalf() for r in real_roots_alg(poly, x)])

    # for comparison, directly numerically compute roots
    print([r for r in poly.nroots() if r.is_real])

    print('\n')
```

Which prints:

```
Poly(sqrt(3)*x**5 + x**3 - 1, x, domain='EX')
[CRootOf(3*x**10 - x**6 + 2*x**3 - 1, 1)]
[0.784991300172933]
[0.784991300172933]


Poly(sqrt(2)*x**5 - sqrt(3)*x**3 + 1, x, domain='EX')
[CRootOf(4*x**20 - 12*x**16 + 9*x**12 - 4*x**10 - 6*x**6 + 1, 0)]
[-1.25721399074364]
[-1.25721399074364]


Poly(x**5 - 3 + CRootOf(y**5 + y**3 - 1, 0)**2, x, domain='EX')
[CRootOf(x**25 - 17*x**20 + 115*x**15 - 387*x**10 + 648*x**5 - 431, 0)]
[1.18109508521431]
[1.18109508521431]


Poly(x**5 + sqrt(3)*x**2 - CRootOf(y**5 + y**3 - 1, 0)**2*x - 1, x, domain='EX')
[CRootOf(x**50 + 4*x**46 - 10*x**45 - 15*x**44 + 6*x**42 - 36*x**41 - 3*x**40 + 120*x**39 + 94*x**38 - 48*x**37 + 84*x**36 + 216*x**35 - 203*x**34 - 568*x**33 - 138*x**32 + 24*x**31 - 584*x**30 - 246*x**29 + 993*x**28 + 924*x**27 + 113*x**26 + 574*x**25 + 741*x**24 - 608*x**23 - 1211*x**22 - 422*x**21 - 617*x**20 - 512*x**19 + 212*x**18 + 732*x**17 + 96*x**16 + 296*x**15 + 135*x**14 - 66*x**13 - 261*x**12 + 142*x**11 - 156*x**10 - 42*x**9 + 109*x**8 - 10*x**7 - 28*x**6 + 40*x**5 - 14*x**4 - 4*x**3 + 6*x**2 - 4*x + 1, 0), CRootOf(x**50 + 4*x**46 - 10*x**45 - 15*x**44 + 6*x**42 - 36*x**41 - 3*x**40 + 120*x**39 + 94*x**38 - 48*x**37 + 84*x**36 + 216*x**35 - 203*x**34 - 568*x**33 - 138*x**32 + 24*x**31 - 584*x**30 - 246*x**29 + 993*x**28 + 924*x**27 + 113*x**26 + 574*x**25 + 741*x**24 - 608*x**23 - 1211*x**22 - 422*x**21 - 617*x**20 - 512*x**19 + 212*x**18 + 732*x**17 + 96*x**16 + 296*x**15 + 135*x**14 - 66*x**13 - 261*x**12 + 142*x**11 - 156*x**10 - 42*x**9 + 109*x**8 - 10*x**7 - 28*x**6 + 40*x**5 - 14*x**4 - 4*x**3 + 6*x**2 - 4*x + 1, 1), CRootOf(x**50 + 4*x**46 - 10*x**45 - 15*x**44 + 6*x**42 - 36*x**41 - 3*x**40 + 120*x**39 + 94*x**38 - 48*x**37 + 84*x**36 + 216*x**35 - 203*x**34 - 568*x**33 - 138*x**32 + 24*x**31 - 584*x**30 - 246*x**29 + 993*x**28 + 924*x**27 + 113*x**26 + 574*x**25 + 741*x**24 - 608*x**23 - 1211*x**22 - 422*x**21 - 617*x**20 - 512*x**19 + 212*x**18 + 732*x**17 + 96*x**16 + 296*x**15 + 135*x**14 - 66*x**13 - 261*x**12 + 142*x**11 - 156*x**10 - 42*x**9 + 109*x**8 - 10*x**7 - 28*x**6 + 40*x**5 - 14*x**4 - 4*x**3 + 6*x**2 - 4*x + 1, 2)]
[-1.16987781989996, -0.615654494012519, 0.828946796272968]
[-1.16987781989996, -0.615654494012519, 0.828946796272968]


Poly(CRootOf(y**5 + y**3 - 3, 0)*x**2 + CRootOf(y**5 + y**3 - 1, 0)*x - 1, x, domain='EX')
[CRootOf(243*x**50 + 486*x**46 + 405*x**45 - 1188*x**44 + 405*x**43 + 1647*x**42 - 675*x**41 - 621*x**40 + 6750*x**39 + 1377*x**38 - 7182*x**37 - 2259*x**36 + 14932*x**35 + 1080*x**34 + 7579*x**33 - 2214*x**32 - 4113*x**31 - 16584*x**30 + 20021*x**29 - 742*x**28 - 2924*x**27 - 625*x**26 + 20623*x**25 + 88*x**24 + 4198*x**23 + 6622*x**22 + 3355*x**21 + 5593*x**20 + 5506*x**19 + 403*x**18 + 2792*x**17 + 86*x**16 + 1907*x**15 + 147*x**14 + 15*x**13 - 101*x**12 - 4*x**11 - 52*x**10 + 25*x**9 - 45*x**8 + 20*x**7 - 29*x**6 + 5*x**5 - 15*x**4 - 5*x**2 - 1, 0), CRootOf(243*x**50 + 486*x**46 + 405*x**45 - 1188*x**44 + 405*x**43 + 1647*x**42 - 675*x**41 - 621*x**40 + 6750*x**39 + 1377*x**38 - 7182*x**37 - 2259*x**36 + 14932*x**35 + 1080*x**34 + 7579*x**33 - 2214*x**32 - 4113*x**31 - 16584*x**30 + 20021*x**29 - 742*x**28 - 2924*x**27 - 625*x**26 + 20623*x**25 + 88*x**24 + 4198*x**23 + 6622*x**22 + 3355*x**21 + 5593*x**20 + 5506*x**19 + 403*x**18 + 2792*x**17 + 86*x**16 + 1907*x**15 + 147*x**14 + 15*x**13 - 101*x**12 - 4*x**11 - 52*x**10 + 25*x**9 - 45*x**8 + 20*x**7 - 29*x**6 + 5*x**5 - 15*x**4 - 5*x**2 - 1, 1)]
[-1.40277952790930, 0.644957328781675]
[-1.40277952790930, 0.644957328781675]
```

I believe the last missing piece is a way to simplify expressions involving algebraic numbers, which we need to check eg that if `r = CRootOf(x**5 + x**3 - 1, 0)` then `r**5 + r**3 - 1` should be identically 0, without float evaluations. Issue #19726 seems to address this by using the minimal polynomial. Would this be the correct approach?

Of course, both solving polynomials with algebraic coefficients and simplifying expressions with algebraic numbers are more general functionality that should perhaps be integrated in a "deeper" level in SymPy. But, I'm not sure what is the most appropriate way. I noticed there was a bunch of discussion on both #23077 and #19726 on this.

oscarbenjamin commented:
> I also did notice that `.count_roots()` fails sometimes for algebraic coefficients.

See gh-26787

> Here is the full implementation with some examples. As you point out, it can be a bit slow especially when the rational lift (is that the right term?) gets quite complicated.

We can make many of the basic operations faster. Let's get this correct first and then focus on what can be made faster afterwards.

> Of course, both solving polynomials with algebraic coefficients and simplifying expressions with algebraic numbers are more general functionality that should perhaps be integrated in a "deeper" level in SymPy. But, I'm not sure what is the most appropriate way.

Expressions with algebraic numbers are automatically simplified in the Poly representation if an algebraic field is used as the domain:
```python
In [23]: Poly(r, y, domain=QQ.algebraic_field(r))
Out[23]: Poly(CRootOf(x**3 - x - 1, 0), y, domain='QQ<CRootOf(x**3 - x - 1, 0)>')

In [24]: Poly(r, y, domain=QQ.algebraic_field(r)) ** 3
Out[24]: Poly(1 + CRootOf(x**3 - x - 1, 0), y, domain='QQ<CRootOf(x**3 - x - 1, 0)>')
```
The bug above that `is_algebraic` is not set on RootOf objects prevents `extension=True` from doing this automatically but that bug is easy to fix and should be fixed regardless.

Solving polynomials with algebraic coefficients should be done by RootOf I think. Ideally the `rootof` function would do this automatically if not RootOf itself. For now we can use a function for this as part of the prototype CAD implementation.

oscarbenjamin commented:
See also gh-26743 and gh-26740.

oscarbenjamin commented:
> The bug above that `is_algebraic` is not set on RootOf objects prevents `extension=True` from doing this automatically but that bug is easy to fix and should be fixed regardless.

I've opened gh-26806 for this.

oscarbenjamin commented:
> > The bug above that `is_algebraic` is not set on RootOf objects prevents `extension=True` from doing this automatically but that bug is easy to fix and should be fixed regardless.
> 
> I've opened [gh-26806](https://github.com/sympy/sympy/pull/26806) for this.

That one is fixed now:
```python
In [1]: r = RootOf(x**3 - x - 1, 0)

In [2]: Poly(r, y, extension=True)
Out[2]: Poly(CRootOf(x**3 - x - 1, 0), y, domain='QQ<CRootOf(x**3 - x - 1, 0)>')
```

oscarbenjamin commented:
> I'm not sure what is going on in `.lift()` either

Fix is in gh-26812

oscarbenjamin commented:
The PRs gh-26816 and gh-26813 fix the last remaining issues that were discussed above.

Once gh-26813 is merged I think that the first step for moving this CAD implementation forward would be to add a function for computing at least the real roots of any polynomial with real algebraic coefficients which could be used by `Poly.real_roots`. I suggest to split that out into a separate PR from this one so that we can get pieces reviewed and merged one step at a time.

You could open a PR based on gh-26813 before it is merged as well and then the two could just be merged together.

mmaaz-git committed:
Attempt to get "simple" sample points

Tries to get a sample point in the following order, if possible: 0, integer, rational. May fail due to numerical issues.

oscarbenjamin commented:
I think some git weirdness happened in the last push...

mmaaz-git commented:
Thanks for the changes. I'm catching up on them today. I ran into some issues trying to get your changes into my branch. Admittedly I'm not very experienced with Git. Is there any way to fix this or should I just make a new branch? 

In the meantime, I will also open the new PR once #26813 is merged as you asked.

oscarbenjamin commented:
> Is there any way to fix this or should I just make a new branch?

It is probably easiest to make a new branch.

> In the meantime, I will also open the new PR once #26813 is merged as you asked.

Hopefully it will be merged soon.

mmaaz-git committed:
Merge branch 'sympy:master' into cad-branch

mmaaz-git committed:
Towards not using floats

Used the new fixes to .lift() to simplify expressions when algebraic numbers are substituted into polynomials. This is now used in lifting as well as when evaluating truth of system after. Also, we use rational approximations to numerical roots if needed. This will soon be remedied!

mmaaz-git commented:
I think I got it working and merged in your changes. I've used the new functionality in #26812 to simplify algebraic expressions by constructing the appropriate algebraic domain. For example, this now works, which did not work before.

```
>>> solve_poly_system_cad([Eq(x**5 + x**3 - 1, 0)], [x])
[{x: CRootOf(x**5 + x**3 - 1, 0)}]
```

As well, we now always use integer/rational sample points for intervals.
```
>>> solve_poly_system_cad([x**5 + x**3 - 1 > 0], [x])
[{x: 2}]
>>> solve_poly_system_cad([x**5 + x**3 - 1 < 0], [x])
[{x: 0}]
```

For the polynomials for which we currently compute numerical roots, I currently wrap them in `Rational()`, just to make everything else work. But of course this will be fixed soon with the other PRs as mentioned.

oscarbenjamin commented:
I think it would be worthwhile to pull small pieces of this out as a separate pull request rather than trying to work on this as one big pull request. It would be worthwhile to go through full review on a smaller PR.

I think that finding at least real roots of a polynomial with real algebraic coefficients is a good candidate for extracting to a separate PR. That needs the `count_roots` fix in gh-26813 but you can build the PR on top of that PR by making a branch on top of that branch.

oscarbenjamin commented:
This test case shows why `nroots` is not sufficient in general:
```python
In [21]: p = expand(((x - sqrt(2))*(x + sqrt(2)))**2) - sqrt(3)/10**100

In [22]: nroots(p)
Out[22]: [-1.4142135623731 + 5.00360173262486e-15⋅ⅈ, -1.4142135623731 - 5.30451210253774e-15⋅ⅈ, 1.4142135623731 + 3.08728310304706e-15⋅ⅈ, 1.4142135623731 - 3.94102083388569e-15⋅ⅈ]
```

mmaaz-git commented:
Sounds good. I'll break the implementation into smaller PRs. I think the best way would be: subresultant and reducta stuff, then the projection operators, then the CAD function itself which brings them together.

And I'll first make a new PR for real roots of a poly with algebraic coefficients.

oscarbenjamin commented:
> subresultant and reducta stuff, then the projection operators, then the CAD function itself which brings them together.

If these are not really independently usable then they probably belong together in a single PR if they are all part of the same algorithm.

oscarbenjamin commented:
Another thing that is independently useful is finding rational sample points for the intervals between the roots of a univariate polynomial. Poly has an `intervals` method but that is not quite what is needed here.

oscarbenjamin commented:
gh-26813 is merged now.

oscarbenjamin commented:
So some of the work here got split out to gh-26945.

Where does that leave us now?

I guess this needs at least some changes but maybe it makes sense to split out some parts again?

Probably a first step is to get to a fully working implementation after gh-26945

It would be good I think to discuss what is an appropriate API. I don't think that "solvers" is the right place for this because this is not what the word "solve" usually means in SymPy.

mmaaz-git commented:
Sorry, been busy with the start of the semester!

I think it makes sense to split out the subresultant and the reducta stuff first. The new subresultant functions are computing the subresultant polynomials and subresultant coefficients. It's the latter, called the PSC (principal subresultant coefficients) that are needed for the projection operator. Both are computed using the subresultant PRS, what SymPy currently calls `subresultants`. I think it makes sense to make these two new public functions, e.g., `subresultant_polys` and `subresultant_coeffs`. For reference, these would line up with the following Mathematica commands:
- `subresultants` -> `SubresultantPolynomialRemainders`
- `subresultant_polys` -> `SubresultantPolynomials`
- `subresultant_coeffs` -> `Subresultants`

As for the reducta stuff (the reductum is defined as the polynomial minus the leading term), I'm not sure if this is of use in other contexts.

Regarding the organization of the code: I think in some module -- perhaps a new one, say `polys.cadtools` (similar to the current `polys.groebnertools`) -- there should be the functions for the Hong projection operator and for computing a CAD adapted to a set of polynomials. I am thinking the CAD function should return something like what your code in #26177 returns: a list of sign conditions and a sample point, over each cell. On that note, I will reuse the code from #26177 to find rational sample points for the intervals. 

There should finally be some sort of "solver" function, if that's the right word, which takes a (Boolean?) expression of polynomial (in)equalities and uses the CAD function to solve it, i.e. satisfying point or indicate none exists. I see that `solvers.polysys` has functions for solving polynomial systems, so maybe this is the right place for such a function?

So basically this is what I see as the roadmap:
- polys.polytools: new functions for subresultant and reducta stuff
- polys.cadtools: Hong projector, sample points, CAD function itself
- polys.solvers: solve a system of polynomial inequalities

What do you think?

oscarbenjamin commented:
> I think it makes sense to make these two new public functions, e.g., `subresultant_polys` and `subresultant_coeffs`. For reference, these would line up with the following Mathematica commands:
> 
> * `subresultants` -> `SubresultantPolynomialRemainders`
> * `subresultant_polys` -> `SubresultantPolynomials`
> * `subresultant_coeffs` -> `Subresultants`

That seems reasonable.

oscarbenjamin commented:
> Sorry, been busy with the start of the semester!

Same here. No rush!

> As for the reducta stuff (the reductum is defined as the polynomial minus the leading term), I'm not sure if this is of use in other contexts.

Possibly not and also an easy enough function anyway:
```python
In [32]: p = x**2 + x + 1

In [33]: p - LT(p)
Out[33]: x + 1
```

> I think in some module -- perhaps a new one, say `polys.cadtools` (similar to the current `polys.groebnertools`) -- there should be the functions for the Hong projection operator and for computing a CAD adapted to a set of polynomials.

Seems reasonable.

> There should finally be some sort of "solver" function, if that's the right word, which takes a (Boolean?) expression of polynomial (in)equalities and uses the CAD function to solve it, i.e. satisfying point or indicate none exists.

I agree. This should be implemented in such a way that it shortcuts once finding a satisfying point though rather than computing a full CAD.

> I see that `solvers.polysys` has functions for solving polynomial systems, so maybe this is the right place for such a function?

No, because this is not usually what the word "solve" means in SymPy although it would be good to have high level functions for this kind of solving like Mathematica's `FindInstance`.

I think that what probably makes sense here is that there should be a object that can represent a system of inequalities and provides methods for transforming e.g. to eliminate some variables or to compute a CAD, find an instance etc. In the first instance though we should just implement functions that do what is needed rather than having a class.
